doc/man/C/xvidcap-dbus-client.1
'\" -*- coding: iso8859-1 -*-
.if \n(.g .ds T< \\FC
.if \n(.g .ds T> \\F[\n[.fam]]
.de URL
\\$2 \(la\\$1\(ra\\$3
..
.if \n(.g .mso www.tmac
.TH xvidcap-dbus-client 1 "23 May 2008" "May 2008" ""
.SH NAME
XVidCap D-BUS Client \- Remote Control Utility for XVidCap
.SH SYNOPSIS
'nh
.fi
.ad l
\fBxvidcap-dbus-client\fR \kx
.if (\nx>(\n(.l/2)) .nr x (\n(.l/5)
'in \n(.iu+\nxu
[--action start|stop|pause ]
'in \n(.iu-\nxu
.ad b
'hy
.SH INTRODUCTION
xvidcap-dbus-client is a small utility to allow for manipulating an
xvidcap capture session from outside xvidcap. Its main purpose is to be used from the command
line or key bindings.
.PP
For inter process communication this utility uses D-BUS. If xvidcap is not running when this
program is invoked to trigger an action, xvidcap is started through D-BUS activation. 
For help, feature requests, or bug-reports please turn to http://www.sourceforge.net/projects/xvidcap/
.SH OPTIONS
This program follows the usual GNU command line syntax, 
with long options starting with two dashes (\*(T<\-\-\*(T>). 
A summary of options is included below.
.TP 
\*(T<\fB\-\-action \fR\*(T>start|stop|pause
This sends a command to xvidcap to either start or stop a capturing session or
to pause or resume one. If xvidcap is not running it will be started.
.SH AUTHORS
xvidcap-dbus-client was written by Karl H. Beckers.
.PP
This manual page was written by Karl H. Beckers <\*(T<karl.h.beckers@gmx.net\*(T>> for 
the xvidcap project.
.PP
This translation was done by TRANSLATOR-CREDITS HERE!!!
.PP
Permission is granted to copy, distribute and/or modify this document under 
the terms of the GNU Free Documentation 
License, Version 1.1 or any later version published by the Free 
Software Foundation; with no Invariant Sections, no Front-Cover 
Texts and no Back-Cover Texts.


ChangeLog

2008-05-26  Karl H. Beckers  <karl.h.beckers@gmx.net>
	src: xtoffmpeg.c, gnome_ui.c, capture.c, 
		xvidcap-dbus-client.c, app_data.h, job.h, 
		gnome_frame.c, gnome_frame.h: fix 
		bug # 1952636
	debian/changelog, debian/rules, Doxyfile, 
		src/xvc_error_item.c, src/gnome-xvidcap.glade,
		doc: man/C/xvidcap-dbus-client.1.xml,
		man/C/xvidcap-dbus-client.1, 
		man/C/xvidcap-dbus-client.1.xml.pot,
		man/C/xvidcap.1.xml, man/C/xvidcap.1,
		man/C/xvidcap.1.xml.pot, man/es/xvidcap.1,
		man/de/xvidcap.1-de.po, 
		man/de/xvidcap-dbus-client.1.xml,
		man/de/xvidcap-dbus-client.1, 
		man/de/xvidcap-dbus-client.1-de.po,
		man/de/xvidcap.1.xml, man/de/xvidcap.1,
		man/it/xvidcap.1-it.po, 
		man/it/xvidcap-dbus-client.1-it.po,
		man/it/xvidcap.1.xml, man/it/xvidcap.1,
		xvidcap/C/xvidcap.xml, xvidcap/C/xvidcap.xml.pot,
		xvidcap/de/xvidcap.xml-de.po,
		xvidcap/de/xvidcap.xml, xvidcap/it/xvidcap.xml-it.po,
		xvidcap/it/xvidcap.xml, po: en.po, ru.po, es.po,
		fr.po, de.po, it.po, zh_CN.po: version bump
	configure.in, aclocal.m4: build host upgrade

2008-05-02  Karl H. Beckers  <karl.h.beckers@gmx.net>
	* src/xtoffmpeg.c, src/gnome_ui.c, src/gnome_frame.c:
		preliminary fix for bug # 1952636 and for
		too small encoding buffer for very small
		areas (though according to ffmpeg it should
		have been large enough)
	* doc/xvidcap/C/figures/menu-show-frame.png: added
		image for new menu item
	* . (properties): updated to current ffmpeg svn

2008-04-17  Karl H. Beckers  <karl.h.beckers@gmx.net>
	* src/xtoffmpeg.c, configure.in, config.h.in, ffmpeg:
		update for current ffmpeg

2008-04-17  Jacques Blouin  <jacdex@gmail.com>
	* po/fr.po: added French translation

2008-03-12  Karl H. Beckers  <karl.h.beckers@gmx.net>
	* src: frame.h, xtoxwd.h, led_meter.h, codecs.h, 
		colors.h, xvidcap-intl.h, dbus-server-object.h, 
		control.h, gnome_warning.h, xtoffmpeg.h, 
		gnome_ui.h, capture.h, app_data.h, job.h, 
		eggtrayicon.h, gnome_frame.h, gnome_options.h, 
		xvc_error_item.h: complete fix for bug # 1885088
		(only _XReadPad and __cplusplus remain which
		are only used here, defined elsewhere)

2008-02-08  Karl H. Beckers  <karl.h.beckers@gmx.net>
	* src: led_meter.h, codecs.h, control.h, 
		gnome_warning.c, app_data.c, app_data.h, 
		job.h, eggtrayicon.h, xv_error_item.c, 
		xvc_error_item.c, Makefile.am, 
		xv_error_item.h, xvc_error_item.h,
		../po: en.po, ru.po, es.po, de.po, it.po, 
		POTFILES.in, zh_CN.po: fix bug # 1885088 
		and also change XvError* to match the 
		XVC_* naming scheme now used elsewhere
	* doc/xvidcap/de: xvidcap.xml-de.po, 
		figures/menu.png, figures/prefs-commands.png, 
		figures/prefs-mf.png, figures/prefs-sf.png, 
		figures/prefs-general.png: docs l10n

2008-02-03  Karl H. Beckers  <karl.h.beckers@gmx.net>
	* src/app_data.c: fix for but #1885105 was
		incomplete
	* src/app_data.c: fix for but #1885105 was
		still flawed

2008-02-02  Karl H. Beckers  <karl.h.beckers@gmx.net>
	* aclocal.m4, src: led_meter.c, codecs.c, 
		colors.c, led_meter.h, codecs.h, 
		colors.h, control.h, xtoffmpeg.c, 
		gnome_ui.c, capture.c, app_data.c, 
		job.c, app_data.h, gnome_frame.c, 
		gnome_options.c, main.c, options.c: 
		indentation and fix for bug # 1880627,
		thanks Markus for the original patch.
	* configure.in, config.h.in, acinclude.m4,
		aclocal.m4: fix for rfe # 1880629
	* doc/xvidcap/it/xvidcap.xml, doc:  
		C/xvidcap-dbus-client.1, C/xvidcap.1, 
		es/xvidcap.1, de/xvidcap-dbus-client.1, 
		de/xvidcap.1, it/xvidcap.1.xml, 
		it/xvidcap.1: regenerated docs

2008-01-14  Karl H. Beckers  <karl.h.beckers@gmx.net>
	* src: gnome_frame.c, gnome_ui.c: fix capture
		type radio button in menu and some
		debug statements.

2008-01-13  Karl H. Beckers  <karl.h.beckers@gmx.net>
	* Makefile.am, Changelog: added EXTRA_DIST
		net.jarre_de_the.Xvidcap.service.in
	* doc/man/it/xvidcap-dbus-client.1-it.po,
		src/options.c, po/en.po, po/ru.po, 
		po/es.po, po/de.po, po/it.po, 
		po/zh_CN.po, doc/man: 
		C/xvidcap-dbus-client.1.xml, 
		C/xvidcap-dbus-client.1.xml.pot, 
		C/xvidcap.1.xml, C/xvidcap.1.xml.pot, 
		de/xvidcap.1-de.po, 
		de/xvidcap-dbus-client.1.xml, 
		de/xvidcap-dbus-client.1, 
		de/xvidcap-dbus-client.1-de.po, 
		de/xvidcap.1.xml, de/xvidcap.1, 
		it/xvidcap.1-it.po: Preparing
		l10n for 1.1.7
	* docs/xvidcap: C/figures/menu-mf.png, 
		C/figures/menu.png, 
		C/figures/menu-cont.png, 
		C/figures/menu-sf.png, 
		C/figures/menu-about.png, 
		C/figures/menu-save-prefs.png, 
		C/figures/prefs-commands.png, 
		C/figures/prefs-mf.png, 
		C/figures/menu-animate.png, 
		C/figures/menu-prefs.png, 
		C/figures/menu-make-video.png, 
		C/figures/menu-quit.png, 
		C/figures/prefs-sf.png, 
		C/figures/prefs-general.png, 
		C/figures/menu-help.png, 
		C/xvidcap.xml, C/xvidcap.xml.pot, 
		de/xvidcap.xml-de.po, de/xvidcap.xml, 
		it/xvidcap.xml-it.po: Preparing
		l10n of manual for 1.1.7

2007-12-16  Benjamin Larsson  <banan@ludd.ltu.se>
	* src: codecs.c, codecs.h: added codecs
		"Flash Screen Video" and "FFmpeg
		Video 1" for lossless screen
		captures

2007-12-16  Karl H. Beckers  <karl.h.beckers@gmx.net>
	* src/xtoffmpeg.c: fix flawed capture size
		calculation with rescale
	* src/gnome_ui.c: fix flawed UI behaviour
		around hiding and showing frame
	* src/Makefile.am: fix build without dbus
 
2007-11-30  Karl H. Beckers  <karl.h.beckers@gmx.net>
	* src: gnome_ui.c, capture.c, app_data.c, 
		app_data.h, gnome_frame.c, gnome_options.c, 
		options.c, gnome-xvidcap.glade: Add
		functionality for following the mouse
		pointer when capturing and hiding 
		the frame.
	* src: codecs.c, codecs.h: add format for 
		direct VOB creation
	* src/app_data.c: fixed array arithmetic for
		error printing in nongui mode

2007-11-23  dAniel hAhler  <daniel@thequod.de>
	* po/de.po: bugfix for missing echo's

2007-11-20  Karl H. Beckers  <karl.h.beckers@gmx.net>
	* src: gnome_ui.c, capture.c, app_data.c, 
		job.c, app_data.h, job.h: make 
		capturing more correct, avoid some
		more potential deadlocks through

src/xvidcap-dbus-client.c
/**
 * \file xvidcap-dbus-client.c
 *
 * This file contains a command line application for remote controlling
 * xvidcap itself through dbus remote function calls.
 *
 */

/*
 * Copyright (C) 2004-07 Karl, Frankfurt
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#ifndef DOXYGEN_SHOULD_SKIP_THIS
#ifdef HAVE_CONFIG_H
#include <config.h>
#endif     // HAVE_CONFIG_H
#endif     // DOXYGEN_SHOULD_SKIP_THIS

#include <getopt.h>
#include <stdlib.h>
#include <stdio.h>
#include <strings.h>
#include <dbus/dbus-glib-bindings.h>

#include "xvidcap-client-bindings.h"
#include "xvidcap-intl.h"

/**
 * \brief enumeration of types of actions supported to be used in a switch
 */
enum ACTION_TYPES
{
    ACTION_START,
    ACTION_STOP,
    ACTION_PAUSE
};

/**
 * \brief displays command line usage
 *
 * @param prog a string containing the name of the program
 */
void
usage (char *prog)
{
    printf (_("Usage: %s, ver %s, khb (c) 2003-07\n"), prog, VERSION);
    printf
        (_
         ("[--action #]      action to perform (\"start\"|\"stop\"|\"pause\")\n"));

    exit (1);
}

/**
 * \brief main function of the application to do the remote function invocation
 *
 * @return completion status
 */

int
main (int argc, char *argv[])
{
    // there's one one argument supported
    struct option options[] = {
        {"action", required_argument, NULL, 0}
    };
    int action = -1;
    int opt_index = 0, c;

    DBusGProxy *proxy;
    DBusGConnection *connection;
    GError *error = NULL;

    while ((c = getopt_long (argc, argv, "v", options, &opt_index)) != -1) {
        switch (c) {
        case 0:                       // it's a long option
            switch (opt_index) {
            case 0:                   // action
                if (strcasecmp (optarg, "start") == 0) {
                    action = ACTION_START;
                } else if (strcasecmp (optarg, "stop") == 0) {
                    action = ACTION_STOP;
                } else if (strcasecmp (optarg, "pause") == 0) {
                    action = ACTION_PAUSE;
                }
                break;
            }
        }
    }
    if (action < 0) {
        usage (argv[0]);
    }

    g_type_init ();

    connection = dbus_g_bus_get (DBUS_BUS_SESSION, &error);
    if (connection == NULL) {
        g_warning ("Unable to connect to dbus: %sn", error->message);
        g_error_free (error);
        // Basically here, there is a problem, since there is no dbus :)
    }
    // This won't trigger activation!
    proxy = dbus_g_proxy_new_for_name (connection,
                                       "net.jarre_de_the.Xvidcap",
                                       "/net/jarre_de_the/Xvidcap",
                                       "net.jarre_de_the.Xvidcap");

    switch (action) {
    case ACTION_START:

        if (!net_jarre_de_the_Xvidcap_start (proxy, &error)) {
            g_warning (_("Could not send start command to xvidcap: %s"),
                       error->message);
            g_error_free (error);
        }
        break;
    case ACTION_STOP:

        if (!net_jarre_de_the_Xvidcap_stop (proxy, &error)) {
            g_warning (_("Could not send stop command to xvidcap: %s"),
                       error->message);
            g_error_free (error);
        }
        break;
    case ACTION_PAUSE:

        if (!net_jarre_de_the_Xvidcap_pause (proxy, &error)) {
            g_warning (_("Could not pause/unpause xvidcap: %s"),
                       error->message);
            g_error_free (error);
        }
        break;
    }

    // Cleanup
    g_object_unref (proxy);

    // The DBusGConnection should never be unreffed, it lives once and is
    // shared amongst the process

    return 0;
}
